File: windows/HarnessPortable.Windows/Models/WorkspaceLayout.cs
namespace HarnessPortable.Windows.Models;

public sealed class WorkspaceLayout
{
    public string Name { get; set; } = "";
    public DateTime SavedAt { get; set; } = DateTime.UtcNow;
    public LayoutNode Root { get; set; } = new();
}

public sealed class LayoutNode
{
    /// <summary>"pane" or "split".</summary>
    public string Kind { get; set; } = "pane";

    /// <summary>"horizontal" or "vertical" for split nodes.</summary>
    public string? Orientation { get; set; }

    public List<LayoutNode> Children { get; set; } = [];

    /// <summary>Tab references for pane nodes, in left-to-right tab order.</summary>
    public List<LayoutTabRef> Tabs { get; set; } = [];

    /// <summary>Serialized AvalonDock DockWidth, e.g. "1|Star" or "320|Pixel".</summary>
    public string? DockWidth { get; set; }

    /// <summary>Serialized AvalonDock DockHeight, e.g. "1|Star" or "240|Pixel".</summary>
    public string? DockHeight { get; set; }
}

public sealed class LayoutTabRef
{
    /// <summary>"tunnel" or "direct".</summary>
    public string Kind { get; set; } = "tunnel";

    public string? ProfileId { get; set; }

    public string? Url { get; set; }

    /// <summary>Optional user-renamed tab label.</summary>
    public string? Label { get; set; }

    /// <summary>
    /// Per-tab dsh web view state (selected session, sidebar fold state and
    /// width, right panel width) as mirrored into the page URL by the
    /// <c>dsh-view-state</c> dsh plugin. Null when the plugin is not installed,
    /// or when the page carried nothing to record.
    /// </summary>
    public LayoutWebState? Web { get; set; }
}

/// <summary>
/// One tab's captured dsh web view state. Only the plugin's three whitelisted
/// parameters are ever read into it — never the page URL itself, which carries
/// dsh's process token on first navigation.
/// </summary>
public sealed class LayoutWebState
{
    /// <summary>Selected session id, absent when there is none.</summary>
    public string? SessionId { get; set; }

    /// <summary>Sidebar width in px; 0 means collapsed.</summary>
    public int? Sidebar { get; set; }

    /// <summary>
    /// Right panel's saved width in px; 0 means no width was recorded yet.
    /// Whether the panel is currently shown is not part of this state.
    /// </summary>
    public int? Rightbar { get; set; }
}

File: macos/HarnessPortable/Models/WorkspaceLayout.swift
import Foundation

enum TabKind: String, Codable, Hashable {
    case management
    case tunnel
    case direct
}

enum LayoutNodeKind: String, Codable, Hashable {
    case pane
    case split
}

enum SplitOrientation: String, Codable, Hashable {
    case horizontal
    case vertical
}

struct WorkspaceTab: Identifiable, Codable, Hashable {
    let id: UUID
    var kind: TabKind
    var profileID: String?
    var url: String?
    var label: String?

    init(
        id: UUID = UUID(),
        kind: TabKind,
        profileID: String? = nil,
        url: String? = nil,
        label: String? = nil
    ) {
        self.id = id
        self.kind = kind
        self.profileID = profileID
        self.url = url
        self.label = label
    }

    private enum CodingKeys: String, CodingKey { case kind, profileID, url, label }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = UUID()
        kind = try container.decode(TabKind.self, forKey: .kind)
        profileID = try container.decodeIfPresent(String.self, forKey: .profileID)
        url = try container.decodeIfPresent(String.self, forKey: .url)
        label = try container.decodeIfPresent(String.self, forKey: .label)
    }
}

struct LayoutNode: Identifiable, Codable, Hashable {
    let id: UUID
    var kind: LayoutNodeKind
    var orientation: SplitOrientation?
    var children: [LayoutNode]
    var tabs: [WorkspaceTab]
    var dockWidth: String?
    var dockHeight: String?

    init(
        id: UUID = UUID(),
        kind: LayoutNodeKind,
        orientation: SplitOrientation? = nil,
        children: [LayoutNode] = [],
        tabs: [WorkspaceTab] = [],
        dockWidth: String? = nil,
        dockHeight: String? = nil
    ) {
        self.id = id
        self.kind = kind
        self.orientation = orientation
        self.children = children
        self.tabs = tabs
        self.dockWidth = dockWidth
        self.dockHeight = dockHeight
    }

    static func pane(_ tabs: [WorkspaceTab] = []) -> LayoutNode {
        LayoutNode(kind: .pane, tabs: tabs)
    }

    static func split(_ orientation: SplitOrientation, _ children: [LayoutNode]) -> LayoutNode {
        LayoutNode(kind: .split, orientation: orientation, children: children, dockWidth: "1|Star", dockHeight: "1|Star")
    }

    private enum CodingKeys: String, CodingKey {
        case kind, orientation, children, tabs, dockWidth, dockHeight
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = UUID()
        kind = try container.decodeIfPresent(LayoutNodeKind.self, forKey: .kind) ?? .pane
        orientation = try container.decodeIfPresent(SplitOrientation.self, forKey: .orientation)
        children = try container.decodeIfPresent([LayoutNode].self, forKey: .children) ?? []
        tabs = try container.decodeIfPresent([WorkspaceTab].self, forKey: .tabs) ?? []
        dockWidth = try container.decodeIfPresent(String.self, forKey: .dockWidth)
        dockHeight = try container.decodeIfPresent(String.self, forKey: .dockHeight)
    }
}

struct WorkspaceLayout: Codable, Identifiable, Hashable {
    var id: String { name }
    var name: String
    var savedAt: String
    var root: LayoutNode

    init(name: String, savedAt: String = ISO8601DateFormatter().string(from: Date()), root: LayoutNode) {
        self.name = name
        self.savedAt = savedAt
        self.root = root
    }
}
